### PR TITLE
Provide ability to customise cljfmt code formatting

### PR DESCRIPTION
**Pull request by iarenaza**
**Describe the solution you'd like**

It should be possible to provide a custom formatting configuration for
cider-nrepl/cljfmt. The cljfmt configuration format is described here:

  https://github.com/weavejester/cljfmt#configuration

The desired solution should have a defcustom variable where the
configuration can be provided.

**Describe alternatives you've considered**

A workaround to this solution is to use an after-save-hook as
described here:

https://github.com/dzer6/cljfmt-graalvm

However, in this case, the two formatters (cider, external cljfmt)
will keep fighting each other causing an awful user experience.

**Additional context**

In https://github.com/clojure-emacs/cider/issues/2907 Bozhidar Batsov
commented:

 > You can just disable dynamic (REPL-powered) indentation if you
 > don't rely on it
 > https://docs.cider.mx/cider/0.26/config/indentation.html
 > Alternative you can completely override the indentation logic of
 > clojure-mode and provide whatever implementation you want. CIDER
 > simply hooks into it and adds new indentation rules there
 > dynamically when connected.

To which Brunno Bonacci responded:

  > However, I still think that this ticket stands. Having the
  > possibility to configure the cljfmt indentation rules is a
  > nice-to-have feature.

**Replace this placeholder text with a summary of the changes in your PR.
The more detailed you are, the better.**

-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [X] The commits are consistent with our [contribution guidelines](../blob/master/.github/CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [X] All tests are passing (`eldev test`)
- [X] All code passes the linter (`eldev lint`) which is based on [`elisp-lint`](https://github.com/gonewest818/elisp-lint) and includes
  - [byte-compilation](https://www.gnu.org/software/emacs/manual/html_node/elisp/Byte-Compilation.html), [`checkdoc`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Tips.html), [check-declare](https://www.gnu.org/software/emacs/manual/html_node/elisp/Declaring-Functions.html), packaging metadata, indentation, and trailing whitespace checks.
- [X] You've updated the [changelog](../blob/master/CHANGELOG.md) (if adding/changing user-visible functionality)
- [ ] You've updated the [user manual](../blob/master/doc) (if adding/changing user-visible functionality)

Thanks!

### CODE REVIEW

File: cider-client.el [hunk 1]
```diff
@@ -210,6 +210,47 @@ buffer, defaults to (cider-current-repl)."
...
+
+  {:indents {org.me/foo [[:inner 0]]}
+   :alias-map {\"me\" \"org.me\"}}
+
+you need to encode it as the following plist:
+
+  '((\"indents\" ((\"org.me/foo\" ((\"inner\" 0))))) (\"alias-map\" ((\"me\" \"org.me\"))))"
+  :type 'list
+  :group 'cider
+  :package-version '(cider . "1.0.0"))
```
Comment (bbatsov) [inline cider-client.el:225]: This should say `1.1.0`.

File: cider-client.el [hunk 2]
```diff
@@ -210,6 +210,47 @@ buffer, defaults to (cider-current-repl)."
 If NS is non-nil, include it in the eval request."
   (nrepl-sync-request:eval input (or connection (cider-current-repl nil 'ensure)) ns))
 
+(defcustom cider-format-code-options nil
+  "A map of options that will be passed to `cljfmt' to format code.
+Assuming this is the Clojure map you want to use as `cljfmt options':
```
Comment (bbatsov) [inline cider-client.el:215]: I think you probably wanted to put quotes only around `cljfmt` here.

File: cider-client.el [hunk 3]
```diff
@@ -712,12 +753,19 @@ The result entries are relative to the classpath."
     (cider-nrepl-send-sync-request)
     (nrepl-dict-get "fn-deps")))
 
-(defun cider-sync-request:format-code (code)
-  "Perform nREPL \"format-code\" op with CODE."
-  (thread-first `("op" "format-code"
-                  "code" ,code)
-    (cider-nrepl-send-sync-request)
-    (nrepl-dict-get "formatted-code")))
+(defun cider-sync-request:format-code (code &optional format-options)
```
Comment (bbatsov) [inline cider-client.el:756]: I think you can just stick the options directly in the request, as this is how we did it in other places. Technically passing this is a param is better, but it's unlikely to ever be used without the options in CIDER.

**Comment by bbatsov** [comment]
Apart from my small remarks the changes look good. Don't forget to mention them in the changlog and the configuration docs.

**Comment by bbatsov** [comment]
@iarenaza ping :-)

**Comment by bbatsov** [comment]
@iarenaza Ping 2 :-)

**Comment by iarenaza** [comment]
@bbatsov Sorry for the delay. Life's got in the way :disappointed: 

I already did the changes you suggested and had added the mention in the changelog. But I was unsure about where to put this in the documentation. There isn't any mention in the documentation at the moment of code formatting via `cljfmt`. So I was doubting between adding it in the "Using CIDER > Misc Features" and "Configuration > Indentation". How do you see it?

**Comment by bbatsov** [comment]
> So I was doubting between adding it in the "Using CIDER > Misc Features" and "Configuration > Indentation". How do you see it?

I'm fine with both, but I'm leaning towards "Misc Features", as it's not directly related to indentation.

**Pull request merged**